### PR TITLE
feat(form): add array collection format in form binding

### PR DESCRIPTION
**Pull request by slowhigh**
Enables the possibility of collecting array data from queries or form data by specifying the collection_format tag in structs.

Several people are waiting for this feature to be updated.

cc @Simon-Saaw

Related PR: https://github.com/gin-gonic/gin/pull/2750
Close https://github.com/gin-gonic/gin/pull/2750

### CODE REVIEW

**Comment by slowhigh** [comment]
Thank you for your comments. @takanuva15

File: binding/form_mapping.go [hunk 1]
```diff
@@ -182,6 +182,34 @@ func trySetCustom(val string, value reflect.Value) (isSet bool, err error) {
...
+		sep = " "
+	case "tsv":
+		sep = "\t"
+	case "pipes":
+		sep = "|"
+	default:
+		return vs, fmt.Errorf("%s is not supported in the collection_format. (csv, ssv, pipes)", cfTag)
+	}
+
+	newVs = make([]string, 0)
```
Comment (appleboy) [inline binding/form_mapping.go:205]:
```go
	totalLength := 0
	for _, v := range vs {
		totalLength += strings.Count(v, sep) + 1
	}
	newVs := make([]string, 0, totalLength)
	for _, v := range vs {
		newVs = append(newVs, strings.Split(v, sep)...)
	}
```

Benchmark

```sh
goos: darwin
goarch: arm64
BenchmarkSplitStringsOriginal-10         1258774               924.5 ns/op
BenchmarkSplitStringsOptimized-10        1716301               697.3 ns/op
PASS
```

**Review by appleboy — CHANGES_REQUESTED** [review]

**Comment by appleboy** [comment]
@slowhigh Any updates?

**Comment by slowhigh** [comment]
@appleboy Thank you. Good idea.

File: binding/form_mapping.go (hunk 1 above)
Comment (slowhigh) [reply binding/form_mapping.go]: I replaced it with the code you mentioned.

Comment (appleboy) [reply binding/form_mapping.go]: @slowhigh Thanks

**Comment by appleboy** [comment]
@slowhigh Lint error https://github.com/gin-gonic/gin/actions/runs/10399149908/job/28847599766?pr=3986

**Comment by slowhigh** [comment]
@appleboy I have addressed the lint error you mentioned.

**Comment by appleboy** [comment]
@slowhigh LGTM

**Review by appleboy — APPROVED** [review]

**Pull request merged**